src/main/java/net/neoforged/neoforge/client/model/data/ModelDataManager.java

```diff
@@ -32,49 +32,123 @@
-                .add(blockEntity.getBlockPos());
-    }
+    public abstract ModelData getAtOrEmpty(BlockPos pos);
+
+    public abstract ModelDataManager.Snapshot snapshotSectionRegion(int sectionMinX, int sectionMinY, int sectionMinZ, int sectionMaxX, int sectionMaxY, int sectionMaxZ);
+
+    public static final class Active extends ModelDataManager {
+        private final Thread owningThread = Thread.currentThread();
+        private final Level level;
+        private final Long2ObjectMap<Set<BlockPos>> needModelDataRefresh = new Long2ObjectOpenHashMap<>();
+        private final Long2ObjectMap<Long2ObjectMap<ModelData>> modelDataCache = new Long2ObjectOpenHashMap<>();
+
+        public Active(Level level) {
+            this.level = level;
+        }
+
+        public void requestRefresh(BlockEntity blockEntity) {
+            if (isOtherThread()) {
+                throw new UnsupportedOperationException("Cannot request ModelData refresh outside the owning thread: " + owningThread);
+            }
+
+            Preconditions.checkNotNull(blockEntity, "BlockEntity must not be null");
+            Preconditions.checkState(blockEntity.getLevel() == level, "BlockEntity does not belong to the level owning this manager");
+            needModelDataRefresh.computeIfAbsent(SectionPos.asLong(blockEntity.getBlockPos()), $ -> new HashSet<>())
+                    .add(blockEntity.getBlockPos());
+        }
 
-    private void refreshAt(ChunkPos chunk) {
-        Set<BlockPos> needUpdate = needModelDataRefresh.remove(chunk);
-
-        if (needUpdate != null) {
-            Map<BlockPos, ModelData> data = modelDataCache.computeIfAbsent(chunk, $ -> new ConcurrentHashMap<>());
-            for (BlockPos pos : needUpdate) {
-                BlockEntity toUpdate = level.getBlockEntity(pos);
-                if (toUpdate != null && !toUpdate.isRemoved()) {
-                    data.put(pos, toUpdate.getModelData());
-                } else {
-                    data.remove(pos);
+        @Override
+        public ModelData getAt(BlockPos pos) {
+            Preconditions.checkArgument(level.isClientSide, "Cannot request model data for server level");
+            long sectionPos = SectionPos.asLong(pos);
+            refreshAt(sectionPos);
+            return modelDataCache.getOrDefault(sectionPos, Long2ObjectMaps.emptyMap()).get(pos.asLong());
+        }
+
+        @Override
+        public ModelData getAtOrEmpty(BlockPos pos) {
+            return Objects.requireNonNullElse(getAt(pos), ModelData.EMPTY);
+        }
+
+        @Override
+        public ModelDataManager.Snapshot snapshotSectionRegion(int sectionMinX, int sectionMinY, int sectionMinZ, int sectionMaxX, int sectionMaxY, int sectionMaxZ) {
+            return new ModelDataManager.Snapshot(this, sectionMinX, sectionMinY, sectionMinZ, sectionMaxX, sectionMaxY, sectionMaxZ);
```

Comment:
I think this should throw if you call it off the main thread. I just did it by accident while porting Embeddium locally.